tools/walletextension/multi_acc_helper.go

```diff
@@ -0,0 +1,169 @@
+	}
+	dataString, ok := data.(string)
+	if !ok {
+		return nil, fmt.Errorf("eth_call request's `data` field was not of the expected type `string`")
+	}
+
+	// We check that the data field is long enough before removing the leading "0x" (1 bytes/2 chars) and the method ID
+	// (4 bytes/8 chars).
+	if len(dataString) < 10 {
+		return nil, nil //nolint:nilnil
+	}
+	dataString = dataString[10:]
+
+	// We split up the arguments in the `data` field.
+	var dataArgs []string
+	for i := 0; i < len(dataString); i += ethCallPaddedArgLen {
+		if i+ethCallPaddedArgLen > len(dataString) {
+			break
+		}
+		dataArgs = append(dataArgs, dataString[i:i+ethCallPaddedArgLen])
+	}
+
+	// We iterate over the arguments, looking for an argument that matches a viewing key address
+	for _, dataArg := range dataArgs {
+		// If the argument doesn't have the correct padding, it's not an address.
+		if !strings.HasPrefix(dataArg, ethCallAddrPadding) {
+			continue
+		}
+
+		maybeAddress := common.HexToAddress(dataArg[len(ethCallAddrPadding):])
+		if _, ok := accClients[maybeAddress]; ok {
+			return &maybeAddress, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no known account found in data bytes")
+}
+
+// Many eth RPC requests provide params as first argument in a json map with similar fields (e.g. a `from` field)
+func parseParams(args []interface{}) (map[string]interface{}, error) {
+	if len(args) == 0 {
+		return nil, fmt.Errorf("no params found to unmarshal")
+	}
+
+	// only interested in trying first arg
+	params, ok := args[0].(map[string]interface{})
+	if !ok {
+		callParamsJSON, ok := args[0].([]byte)
+		if !ok {
+			return nil, fmt.Errorf("first arg was not a byte array")
+		}
+
+		err := json.Unmarshal(callParamsJSON, &params)
+		if err != nil {
+			return nil, fmt.Errorf("first arg couldn't be unmarshalled into a params map")
+		}
+	}
+
+	return params, nil
+}
+
+// proxyRequest will try to identify the correct EncRPCClient to proxy the request to the Obscuro node, or it will attempt
+//
+//	the request with all clients until it succeeds
+func proxyRequest(rpcReq *rpcRequest, rpcResp interface{}, accClients map[common.Address]*rpcclientlib.EncRPCClient) error {
+	// for obscuro RPC requests it is important we know the sender account for the viewing key encryption/decryption
+	suggestedClient := suggestAccountClient(rpcReq, accClients)
+
+	var err error
+	if suggestedClient != nil {
+		// todo: if we have a suggested client, should we still loop through the other clients if it fails?
+		// 		The call data guessing won't often be wrong but there could be edge-cases there
+		err = executeCall(suggestedClient, rpcReq, &rpcResp)
+	} else {
+		// we attempt the request with every client until we have a successful execution
+		log.Info("appropriate client not found, attempting request with up to %d clients", len(accClients))
+		for _, client := range accClients {
+			err = executeCall(client, rpcReq, &rpcResp)
```

Comment:
one last thing.
executeCall should prob be in the same file.

Reply:
Oops, yeah good spot :)